### PR TITLE
[RHPAM-4438] Sensitive Information In ENV And Pod Logs

### PR DESCRIPTION
**Pull request by davidesalerno**
Signed-off-by: Davide Salerno <dsalerno@redhat.com>

See: [RHPAM-4438](https://issues.redhat.com/browse/RHPAM-4438)

In order to avoid to print sensitive information in the logs in a production environment, it is better to log certain type of information only when the logger is not at DEBUG level (usually in a Prod env log level is at Info)

### CODE REVIEW

**Comment by davidesalerno** [comment]
@tchughesiv Can you help me reviewing this change?

File: go.mod [hunk 1]
```diff
@@ -14,7 +14,7 @@ require (
 	github.com/pkg/errors v0.9.1
 	github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring v0.50.0
 	github.com/stretchr/testify v1.7.0
-	go.uber.org/zap v1.19.0
+	go.uber.org/zap v1.21.0
```
Comment (ruromero) [inline go.mod]: Why is this needed?

**Review by ruromero — COMMENTED** [review]
Thanks for the PR, I only have one question, otherwise LGTM

File: pkg/resource/compare/defaults.go [hunk 2]
```diff
@@ -176,7 +176,7 @@ func equalDeploymentConfigs(deployed client.Object, requested client.Object) boo
 	pairs = append(pairs, [2]interface{}{dc1.Spec, dc2.Spec})
 	equal := EqualPairs(pairs)
 	if !equal {
-		logger.Info("Resources are not equal", "deployed", deployed, "requested", requested)
+		logger.Debugf("Resources are not equal", "deployed", deployed, "requested", requested)
```
Comment (jakubschwan) [inline pkg/resource/compare/defaults.go]: For case when resources or objects are not equal. Can we keep there at least some info output. Otherwise it might be hard for users to spot what is wrong.
If the debug output is not allowed, it would be good to print some info log (or warning) as well. So it can be debug easier. I think it should be okay to just print message like "Resources are not equal. For more details use debug option for your Operator."

File: go.mod (hunk 1 above)
Comment (davidesalerno) [reply go.mod]: @ruromero This change is not strictly needed for this fix and it is due to the attempt to add some unit test using zaptest (go.uber.org/zap/zaptest) which would require huge refactoring and for this reason at the moment they are not in this PR.
I'll revert it.

File: pkg/resource/compare/defaults.go (hunk 2 above)
Comment (davidesalerno) [reply pkg/resource/compare/defaults.go]: @jakubschwan thank you for the suggestion, I agree with your concern and I'll do the change

File: pkg/resource/compare/defaults.go [hunk 3]
```diff
@@ -643,7 +649,8 @@ func EqualPairs(objects [][2]interface{}) bool {
 func Equals(deployed interface{}, requested interface{}) bool {
 	equal := reflect.DeepEqual(deployed, requested)
 	if !equal {
-		logger.Info("Objects are not equal", "deployed", deployed, "requested", requested)
+		logger.Info("Objects are not equal. For more details set the Operato log level at DEBUG.")
```
Comment (ruromero) [inline pkg/resource/compare/defaults.go]:
```suggestion
		logger.Info("Objects are not equal. For more details set the Operator log level to DEBUG.")
```

File: pkg/resource/compare/defaults.go [hunk 4]
```diff
@@ -615,7 +620,8 @@ func equalBuildConfigs(deployed client.Object, requested client.Object) bool {
 	pairs = append(pairs, [2]interface{}{bc1.Spec, bc2.Spec})
 	equal := EqualPairs(pairs)
 	if !equal {
-		logger.Info("Resources are not equal", "deployed", deployed, "requested", requested)
+		logger.Info("Resources are not equal. For more details set the Operato log level at DEBUG.")
```
Comment (ruromero) [inline pkg/resource/compare/defaults.go]:
```suggestion
		logger.Info("Resources are not equal. For more details set the Operator log level to DEBUG.")
```

File: pkg/resource/compare/defaults.go [hunk 5]
```diff
@@ -542,7 +546,8 @@ func equalSecrets(deployed client.Object, requested client.Object) bool {
 	pairs = append(pairs, [2]interface{}{secret1.Data, secret2.Data})
 	equal := EqualPairs(pairs)
 	if !equal {
-		logger.Info("Resources are not equal", "deployed", deployed, "requested", requested)
+		logger.Info("Resources are not equal. For more details set the Operato log level at DEBUG.")
```
Comment (ruromero) [inline pkg/resource/compare/defaults.go]:
```suggestion
		logger.Info("Resources are not equal. For more details set the Operator log level to DEBUG.")
```

**Review by ruromero — CHANGES_REQUESTED** [review]
I'd rather use a conditional logging instead of logging twice. Something like this:
```
		if logs.IsDebugLevel() {
			logger.Debugf("Resources are not equal", "deployed", deployed, "requested", requested)
		} else {
			logger.Info("Resources are not equal. For more details set the Operator log level to DEBUG.")
		}
```

I think this function should be enough:

```
// IsDebugLevel returns whether the current log level is Debug
func IsDebugLevel() bool {
	return GetBoolEnv(DebugTrue.Name)
}
```

**Review by ruromero — APPROVED** [review]
LGTM

**Comment by davidesalerno** [comment]
@ruromero Thanks for the good advices, I integrated it and it is much better than before

**Comment by brusdev** [comment]
@davidesalerno @ruromero this change is causing #110. I created the PR #111 to fix #110 partially reverting [[RHPAM-4438] Sensitive Information In ENV And Pod Log](https://github.com/RHsyseng/operator-utils/commit/7a9086dae039b0dc22aaa8a64462f036bdc770a1). Could you review it?
